src/placeos-driver/interface/sensor.cr

```diff
@@ -0,0 +1,137 @@
+require "json"
+require "s2_cells"
+
+module PlaceOS::Driver::Interface::Sensor
+  enum SensorType
+    Temperature # https://en.wikipedia.org/wiki/Celsius
+    Humidity    # https://en.wikipedia.org/wiki/Humidity#Relative_humidity (percentage)
+    Illuminance # https://en.wikipedia.org/wiki/Illuminance (lux)
+    Pressure    # https://en.wikipedia.org/wiki/Pascal_(unit)
+    Trigger     # Proximity or motion etc
+    Switch      # On or Off - water leak, door open etc
+    Level       # Percentage + raw value + technical human readable (battery, fuel, water tank, oxygen, Co2)
+    Flow        # https://en.wikipedia.org/wiki/Cubic_metre_per_second (gas or liquid)
+    Counter
+    Acceleration # MetrePerSecondSquared
+    Speed        # MetrePerSecond
+    # Gyroscope: roll, pitch, yaw - Angle
+    Roll
+    Pitch
+    Yaw
+    Compass              # direction (magnetic field)
+    Current              # https://en.wikipedia.org/wiki/Ampere
+    Voltage              # https://en.wikipedia.org/wiki/Volt
+    ElectricalResistance # https://en.wikipedia.org/wiki/Ohm
+    Radiation            # https://en.wikipedia.org/wiki/Sievert
+    Distance
+  end
+
+  # Using SI units and SI derived units
+  enum Unit
+    Celsius               # Temperature - celsius over Kelvin to avoid some conversions
+    Percentage            # Humidity, Level
+    Lux                   # Illuminance
+    Pascal                # Pressure
+    Boolean               # Trigger, Switch
+    Integer               # Counter
+    CubicMetrePerSecond   # Liquid or gas flow rate
+    MetrePerSecondSquared # Acceleration
+    MetrePerSecond        # Speed
+    Angle                 # Compass, Accel, Gyro
+    Ampere                # Current
+    Volt                  # Voltage
+    Ohm                   # ElectricalResistance
+    Sievert               # Radiation
+    Metre                 # Distance
+  end
```

Comment:
There is a lot of potential for this to either a) explode over time, or b) limit a number of use cases by enforcing this at the interface level.

An alternative way of viewing this may be to encode the kind of measure that the sensor provides, which can then influence how it is modelled. A incomplete example of this would be:
1. One of a number of discrete states, with these states being generic
2. Boolean (a degenerate case of 1, but likely useful)
3. Integer
4. Bounded integer
5. Non-negative integer (degenerate of 4)
6. {3, 4, 5}, but `s/integer/float/`
7. etc

Units can the be metadata associated with this, rather than intrinsic to the type.

_If_ there is a need to strictly encode units, there is a well defined set of base SI units for the fundamental quantities (length, mass, time, electric current, temperature, luminous intensity and the amount of substance), which can then be used to build derived units with higher dimensionality. That like goes _way_ to deep away from what's actually useful here though.

Reply:
So the idea here is to limit the number of use cases.
A simple format that is easy to put on a map or associate with a room.

i.e. you have temperate sensor that is positioned to a location on the floor
we can then collect all the temperature sensors in an area, average them and that's the temperature for that area
similarly with humidity and Co2 levels etc

it both makes it simple for the backend to process, the frontend to display and the analytics has the raw sensor data, the processed data and the area summaries

Everything is being normalised to SI or Official SI derived units - selecting derived units where it makes more sense
like Celsius over Kelvin as no one knows Kelvin and the frontend only needs to transform for the USA

Reply:
That makes sense.

In that case then there is duplicate information across what here in `Unit` and what is in `SensorType`. For example, using this model, `Unit::Celsius` implies `SensorType::Temperature`. Equally, this also leaves things open to receiving an `Illuminance` point with a value in `CubicMetersPerSecond`.

What about shuffling this so that sensor types have a defined unit, but then also support mapping from other compatible units in the setter?

Reply:
good point